fuse/nodefs/fsconnector.go

```diff
@@ -126,9 +135,23 @@ func (c *FileSystemConnector) forgetUpdate(nodeID uint64, forgetCount int) {
 
 	if forgotten, handled := c.inodeMap.Forget(nodeID, forgetCount); forgotten {
 		node := (*Inode)(unsafe.Pointer(handled))
+		if node.IsDir() && len(node.Children()) > 0 {
+			// This directory node has children that have not yet been forgotten.
+			// We don't want to remove it from the tree (i.e. remove it from its parent)
+			// because the children may have parent pointers to it.
+			// A later LOOKUP will return this node with a new NodeID.
+			return
+		}
 		node.mount.treeLock.Lock()
-		c.recursiveConsiderDropInode(node)
+		// Remove Inode from all parents
```

Comment:
I don't agree with this: this conflates two things, which is: 1) the inode being known to the kernel, and 2) whether or not the the inode should be removed from the tree.  1) is a mattter of erasing the ID from the map, and 2) is a type of garbage collection. 

The clientInodes support is completely in pathfs right now, and I don't see why the support for it has to migrate into nodefs. Can you verify that libfuse needs to modify the lowlevel support code to support hardlinks? If not, we should not need to either.

All of the comments you added to existing code seem fine to me. If you can prepare a separate commit that just adds those, I can merge that without any issues.